version: '3.9'

services:
  keycloak:
    image: quay.io/keycloak/keycloak:24.0
    container_name: keycloak
    environment:
      KEYCLOAK_ADMIN: admin
      KEYCLOAK_ADMIN_PASSWORD: admin
      KC_HEALTH_ENABLED: true
      KC_LOG_LEVEL: debug
      DB_VENDOR: mysql
      DB_ADDR: keycloak-db
      DB_DATABASE: keycloak
      DB_USER: keycloak
      DB_PASSWORD: keycloakpassword
    healthcheck:
      test:   [
        "CMD-SHELL",
        'exec 3<>/dev/tcp/localhost/8080; echo -e "GET /health/ready HTTP/1.1\nhost: localhost:8080\n" >&3; timeout --preserve-status 1 cat <&3 | grep -m 1 status | grep -m 1 UP; ERROR=$?; exec 3<&-; exec 3>&-; exit $ERROR'
      ]
      interval: 5s
      timeout: 5s
      retries: 10
    command: ["start-dev"]
    volumes:
      - keycloak_data:/opt/jboss/keycloak/standalone/data
    ports:
      - "8083:8080"
      - "8084:8443"
    networks:
      - local_network
    depends_on:
      keycloak-db:
        condition: service_healthy
  keycloak-db:
    image: 'mysql:latest'
    container_name: keycloak-db
    environment:
      MYSQL_DATABASE: keycloak
      MYSQL_USER: keycloak
      MYSQL_PASSWORD: keycloakpassword
      MYSQL_ROOT_PASSWORD: ':u0=0TI^3@lI'
    ports:
      - '3306'
    volumes:
      - keycloak_db_data:/var/lib/mysql
#      - './mysql/init-scripts:/docker-entrypoint-initdb.d'
    healthcheck:
      test: ["CMD", "mysqladmin", "ping", "-h", "localhost"]
      interval: '10s'
      timeout: '5s'
      retries: 3
    networks:
      - local_network
  traccar:
    image: traccar/traccar:6.2
    container_name: traccar
    environment:
      CONFIG_USE_ENVIRONMENT_VARIABLES: "true"
      DATABASE_DRIVER: "com.mysql.cj.jdbc.Driver"
      DATABASE_URL: "jdbc:mysql://traccar-db:3306/traccar?allowPublicKeyRetrieval=true&useSSL=false&allowMultiQueries=true&autoReconnect=true&useUnicode=yes&characterEncoding=UTF-8&sessionVariables=sql_mode=''"
      DATABASE_USER: traccar
      DATABASE_PASSWORD: traccarpassword
      WEB_URL: "http://localhost:8082"
      WEB_DEBUG: "true"
      LOGGER_CONSOLE: "true"
      LOGGER_FULLSTACKTRACES: "true"
      OPENID_FORCE: "false"
      OPENID_CLIENT_ID: iotwatch-traccar
      OPENID_CLIENT_SECRET: 7AJ4FkVNj7JwjU8mPp63SZP11RvaxViG
      OPENID_ISSUER_URL: http://keycloak:8080/realms/iotwatch
    ports:
      - "8082:8082"
      - "5000-5245:5000-5245"
      - "5000-5245:5000-5245/udp"
    volumes:
      - ./traccar/logs:/opt/traccar/logs:rw
      - traccar_config:/opt/traccar/conf
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
    depends_on:
      traccar-db:
        condition: service_healthy
      keycloak:
        condition: service_healthy
    networks:
      - local_network
  traccar-db:
    image: 'mysql:latest'
    container_name: traccar-db
    environment:
      MYSQL_DATABASE: traccar
      MYSQL_USER: traccar
      MYSQL_PASSWORD: traccarpassword
      MYSQL_ROOT_PASSWORD: ':u0=0TI^3@lI'
    ports:
      - '3306'
    volumes:
      - traccar_db_data:/var/lib/mysql
    #      - './mysql/init-scripts:/docker-entrypoint-initdb.d'
    healthcheck:
      test: ["CMD", "mysqladmin", "ping", "-h", "localhost"]
      interval: '10s'
      timeout: '5s'
      retries: 3
    networks:
      - local_network

volumes:
  keycloak_data:
  traccar_config:
  keycloak_db_data:
  traccar_db_data:
networks:
  local_network:
    driver: bridge